### ISSUE TITLE
Single GPU support?

### ISSUE THREAD

**Issue by f0rmatme** [HikariKnight/quickpassthrough#9]
Does this have support built in if you don't have two gpus and just want to pass through one?

**Comment by HikariKnight** [comment]
single gpu (no integrated gpu on the cpu) requires a lot more work (and the setup is very different), but it is something i want this script to be able to help configure.

For multi gpu you reserve one of the gpus for the VM to use, preventing linux from using it, easy and simple to setup and removes a lot of complexity (as long as the 2 cards are very different).

While for single gpu, the VM software you use, has to support hook scripts to run before a VM starts and after the VM has stopped and released its resources so that you can hot-plug the gpu, this also brings a host of other issues in that you might need to patch the VBIOS or Get your cards VBIOS from techpowerup (as i noticed dumping the VBIOS in linux does not always produce a useable VBIOS file for passthrough for all cards.

It also does not help that the hook scripts are different depending on what you run on your own desktop, as an example here is mine for plugging my host card (i test both multi and singlegpu passthrough on my system) to a VM, then there is another script that is essentially the below script in reverse in order to hook it back to the host.

```bash
#!/bin/bash
# Helpful to read output when debugging
set -x

source /etc/libvirt/hooks/vfio.conf

# Stop display manager
systemctl stop display-manager.service
## Uncomment the following line if you use GDM and/or a window manager (replace awesome with your WM)
killall gdm-x-session
killall -9 awesome

# Replace hk with your username
pulse_pid=$(pgrep -u hk pulseaudio)
pipewire_pid=$(pgrep -u hk pipewire-media)
kill $pulse_pid
kill $pipewire_pid


# Unbind VTconsoles (change these to match your vtconsoles)
echo 0 > /sys/class/vtconsole/vtcon0/bind
echo 0 > /sys/class/vtconsole/vtcon1/bind

# Unbind EFI-Framebuffer (uncomment if you use an nvidia card)
#echo efi-framebuffer.0 > /sys/bus/platform/drivers/efi-framebuffer/unbind

# Avoid a Race condition by waiting 2 seconds. This can be calibrated to be shorter or longer if required for your system
sleep 5

# Unload amdgpu
modprobe -r amdgpu

# Unload nvidia driver (uncomment if you are passing through an nvidia card)
#modprobe -r nvidia
#modprobe -r nvidia_drm
#modprobe -r gpu_sched
#modprobe -r ttm
#modprobe -r drm_kms_helper
#modprobe -r i2c_algo_bit
#modprobe -r drm
#modprobe -r snd_hda_intel

# Unbind the GPU from display driver (replace 0000_42_00_1 and 2 with your PCI IDs)
virsh nodedev-detach pci_0000_$DEVICE_VIDEO
virsh nodedev-detach pci_0000_$DEVICE_AUDIO

# Load VFIO Kernel Module  
modprobe vfio
modprobe vfio_pci
modprobe vfio_iommu_type1 
```

it is still a work in progress but safe to say it requires a lot more fiddling from the user as there is another script that runs before the above one to make sure certain applications are properly closed before killing the desktop environment and display manager (in order to avoid other issues)

To my knowledge only libvirt (virt-manager) supports hook scripts that at the moment, Quickemu has no support for it yet.

The TLDR;
Single gpu passthrough is a lot more complex with more nuanced, support is not planned anytime soon, but it is planned for the future.

**Comment by Magnetuz** [comment]
Hello @HikariKnight.

I would like take the chance to make same questions. Sorry if this is redundant.

I didn't understand how does it work in the case of a dedicated GPU and an integrated GPU (on the CPU). Will it reserve de dedicated GPU to the guest VM and the integrated GPU to the host OS or is it configurable? When does this reservation start? What happens in case some application is already using the dedicated GPU when the VM starts? And what happens after the VM application is terminated? Is the dedicated GPU available again to the host OS?

Thank you.

**Comment by HikariKnight** [comment]
> Hello @HikariKnight.
> 
> I would like take the chance to make same questions. Sorry if this is redundant.
> 
> I didn't understand how does it work in the case of a dedicated GPU and an integrated GPU (on the CPU). Will it reserve de dedicated GPU to the guest VM and the integrated GPU to the host OS or is it configurable? When does this reservation start? What happens in case some application is already using the dedicated GPU when the VM starts? And what happens after the VM application is terminated? Is the dedicated GPU available again to the host OS?
> 
> Thank you.

depending on how you configure it.
but the iGPU will be reserved for the host no matter what. (it will always be too closely tied to other stuff as its inside the cpu so it will be in the same IOMMU group as other things or with a bridge which will drag in more things from other groups.)

this means that the host need to have the monitors connected to the iGPU if they want to go the "single gpu" approach that way and the dedicated GPU will always not be usable in the host unless you make a script to free up the GPU on shutdown and then switch back the cables and reconfigure your monitor settings each time you shut off the VM, something which is very tedious

When we say a gpu gets reserved for the guest, what we mean is that we prevent the usual driver from binding to it, something you can only do if there are no graphical applications running, hence why we set the active driver for the guest GPU to vfio-pci as the GPU will then effectively be idle with nothing running on it and free to be used by vfio in qemu.
If you want to unbind this driver you might be able to do that while a graphical environment is running but you will not be able to rebind it to vfio-pci without shutting down you display manager and graphical environment. Shutting down the display manager will also force close all your applications and the rest of your session. Again not a good solution

so the way we do it with 2 gpus (and igpu+dgpu) is that we bind 1 gpu to vfio-pci at boot, before the gpu driver loads, and leave it at that. i decided that to make it easy enough to "disable" people can just go into their bootloader menu, press E on the boot entry you use and remove the vfio parameters and boot it like that when they do not need the gpu reserved to the VM.

I have after careful consideration decided to scrap "single SPU" support though, at least the kind where you use the gpu on the host when not using the VM, this is due to the complexity involved to configure it on peoples systems and getting it done correctly.

Single GPU support in "headless" machines will still work like it does now where the GPU will be reserved for VM use only and you rely on ssh or a reboot (to remove the kernel parameters) to get access to the tty of the server. This is a more supported single GPU usecase due to it being no different from a dual GPU setup.

**Comment by Magnetuz** [comment]
Hi @HikariKnight.

Thank you for your reply. It was very clarifying.

Regarding your scripts, the readme says they modify the bootloader with the proper configuration for the passthrough. My question is: is it creating a new boot configuration that could be selected on the boot screen (like when using dual boot with Linux and Windows) or just modifying the default boot? 

In case this selectable option is not the answer, do you think it is complicated to configure this?

Best regards.

**Comment by HikariKnight** [comment]
hi @Magnetuz 
it modifies the default.
editing grub is a pain in general code wise, using grubby is ok and systemd boot is pretty easy with kernelstub
however kernelstub does not allow setting kernel parameters for individual boot entries, hence why i just edit the default (as kernelstub is what i use on my system and test for while i verify the other 2 in VMs)

it is still pretty easy to go into the bootloader menu and just edit out the kernel parameters for a singular boot (because if you were to use the gpu you would have to physically move display cables anyway)
Plus the default kernel parameters also get applied to newer kernels automatically which is the expected behavior.

**Comment by Magnetuz** [comment]
Hi @HikariKnight.

Thank you for the explanation.

In the end, the idea is to have the passthrough as a "pseudo-static" configuration rather than a dynamic hot plug feature, as one need to change the display cables and patch the bootloader each time the passthrough is enabled/disabled. Is it correct?

**Comment by HikariKnight** [comment]
> Hi @HikariKnight.
> 
> Thank you for the explanation.
> 
> In the end, the idea is to have the passthrough as a "pseudo-static" configuration rather than a dynamic hot plug feature, as one need to change the display cables and patch the bootloader each time the passthrough is enabled/disabled. Is it correct?

this is correct in any case where 2 gpus are involved (this includes igpu with dgpu too)

with a single dgpu and no igpu you would have to kill the graphical session each time you switch and use hook script tailored to your system to make sure everything graphical is closed and your gpu is freed from any task before unbinding the driver and then binding it to vfio-pci and then doing everything in reverse when the VM shuts down and there are quirks related to which WM you use, what DE you use, do you use a display manager or not? if so which one or are you using an obscure one?

Also the hook scripts are per VM.

This will lead to several things if you do use true Single GPU passthrough and start a VM
* All graphical applications will be force closed when booting the VM (session will not be saved and data will be lost)
* You will have a black screen while the hook scripts run and if you get any issues, well you got no display at that point.
* Depending on the GPU you may not get visual access to the VM bios/uefi
* Depending on the guest OS a reboot might be needed after the VM is shut down anyway (this is the case for MacOS and with very few GPUs when using windows 11 as the guest OS)
* Some gpus are still affected by the PCI reset but and is not covered by the vendor-reset kernel module
* Once the VM is finally booted and there is already cables to your monitor attached to the dGPU, you will be able to see the screen on your monitors (meaning its the closest you get to a hotplug function)
